### PR TITLE
Add the SpanData class, for future use in export

### PR DESCRIPTION
**Pull request by jkwatson**

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+  }
+
+  /**
+   * An immutable implementation of {@link io.opentelemetry.trace.Link}.
+   *
+   * @since 0.1.0
+   */
+  @Immutable
+  @AutoValue
+  public abstract static class Link implements io.opentelemetry.trace.Link {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: This is not needed, use from the API `util.Links.create`

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+  }
+
+  /**
+   * A timed event representation.
+   *
+   * @since 0.1.0
+   */
+  @Immutable
+  @AutoValue
+  public abstract static class TimedEvent {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: Change TimedEvent to:
* timestamp
* name
* attributes

And remove the Event from the SpanData. A TimedEvent can be constructed from:
* timestamp + api.Event
* timestamp + name
* timestamp + name + attributes

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+   * number of seconds since the Unix Epoch.
+   *
+   * <p>Defined here instead of using {@code Instant} because the API needs to be Java 1.7
+   * compatible.
+   *
+   * @since 0.1.0
+   */
+  @Immutable
+  @AutoValue
+  public abstract static class Timestamp {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: Already defined in the API.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 1 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: Agreed. I had been planning on cleaning all of this up after the fact, but we can push that up front.  (we had just copied @carlosalberto 's `SpanData` class as the starting point)

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 2 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: done

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: we're adding javadoc now

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+    /**
+     * Returns the {@code Timestamp} of this event.
+     *
+     * @return the {@code Timestamp} of this event.
+     * @since 0.1.0
+     */
+    public abstract Timestamp getTimestamp();
+
+    public abstract String getName();
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: Please add some javadoc.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+
+    public abstract Map<String, AttributeValue> getAttributes();
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: ditto.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+   * @since 0.1.0
+   */
+  public abstract Timestamp getEndTimestamp();
+
+  /**
+   * The trace id for this span, retrieved from the associated SpanContext.
+   *
+   * @return the TraceId
+   */
+  public TraceId getTraceId() {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: Consider to either remove the getContext or the individual things. May be better to just keep all the getters here and remove the getSpanContext.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 4 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: done

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 5 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: done

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 6 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: how about making the getContext() method package-access, and expose the pieces as getters that delegate? Then, we don't need to add 4 new fields to the `SpanData` class itself to replace the context.

Comment (breedx-nr) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: I went ahead and removed the context and put the 4 fields in its place.  Feels like the thing has grown a bit, but it's fine.  👍

**Comment by breedx-nr** [comment]
This PR is ready for another 👀

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    public abstract Builder resource(Resource resource);
+
+    /**
+     * Set the name of the span. Must not be null.
+     *
+     * @param name the name.
+     * @return this
+     * @since 0.1.0
+     */
+    public abstract Builder name(String name);
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: for Builder methods we should use setName?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder traceId(TraceId traceId);
+
+    public abstract Builder spanId(SpanId spanId);
+
+    public abstract Builder traceFlags(TraceFlags traceFlags);
+
+    public abstract Builder tracestate(Tracestate tracestate);
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: javadocs?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+   * @param startTimestamp the start {@code Timestamp} of the {@code Span}.
+   * @param attributes the attributes associated with the {@code Span}.
+   * @param timedEvents the events associated with the {@code Span}.
+   * @param links the links associated with the {@code Span}.
+   * @param status the {@code Status} of the {@code Span}.
+   * @param endTimestamp the end {@code Timestamp} of the {@code Span}.
+   * @return a new immutable {@code SpanData}.
+   * @since 0.1.0
+   */
+  public static SpanData create(
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: Why create if we have a builder? just remove this.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 7 above)
Comment (breedx-nr) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: That doesn't seem to be the current standard.  My preference would be to omit `set`.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 8 above)
Comment (breedx-nr) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: Whoops!  Will add.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 9 above)
Comment (breedx-nr) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: There's a little bit of helpful destructuring of the context and some defaulting/protecting of the attributes, links, and events collections.  Seems like there's enough value in there that we shouldn't let it go.

My instinct would be to put move that `create()` to be the `build()` method of the `Builder` but AutoValue makes that challenging doesn't it?  Thoughts?

Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: For the immutable copies see https://github.com/google/auto/blob/master/value/userguide/builders-howto.md#-normalize-modify-a-property-value-at-build-time. Use autoBuild() and write your own build.

For the SpanContext you can have an implemented setter on the Builder that calls setTraceId/set... Not sure if that is needed, in my opinion that can be done by the caller.

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: cool. we'll make it so. I didn't know about that technique (I'm admittedly pretty new to AutoValue)

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java (hunk 7 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: let's try to stick with set please. We use set consistently in this repo, if you want to make a change for that we should do it consistently on the entire repo

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: 👍

File: sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+  }
+
+  /**
+   * Creates a new Builder for creating an SpanData instance.
+   *
+   * @return a new Builder.
+   * @since 0.1.0
+   */
+  public static Builder newBuilder() {
+    return new AutoValue_SpanData.Builder();
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: We should initialize the default values for non-nullable members, with empty list/maps or default values.

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java]: yep, agreed. fix incoming

File: sdk/src/test/java/io/opentelemetry/sdk/trace/export/SpanDataTest.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+import io.opentelemetry.trace.util.Events;
+import io.opentelemetry.trace.util.Links;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class SpanDataTest {
```
Comment (bogdandrutu) [inline sdk/src/test/java/io/opentelemetry/sdk/trace/export/SpanDataTest.java]:
```
/** Unit tests for {@link SpanData}. */
@RunWith(JUnit4.class)
```

Comment (breedx-nr) [reply sdk/src/test/java/io/opentelemetry/sdk/trace/export/SpanDataTest.java]: 👍 added this.

**Comment by bogdandrutu** [comment]
@breedx-nr @jkwatson only one item left to move it from exporter to trace.

**Comment by jkwatson** [comment]
I'll make this change first thing in the morning, over coffee. Thanks for all the great feedback!

**Comment by jkwatson** [comment]
ok, SpanData has been moved up to the trace package

File: sdk/src/main/java/io/opentelemetry/sdk/trace/SpanData.java [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+     * Create a new SpanData instance from the data in this.
+     *
+     * @return a new SpanData instance
+     * @since 0.1.0
+     */
+    public SpanData build() {
+      // make unmodifiable copies of any collections
+      setAttributes(Collections.unmodifiableMap(new HashMap<>(getAttributes())));
+      setTimedEvents(Collections.unmodifiableList(new ArrayList<>(getTimedEvents())));
+      setLinks(Collections.unmodifiableList(getLinks()));
```
Comment (songy23) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/SpanData.java]:
```suggestion
      setLinks(Collections.unmodifiableList(new ArrayList<>(getLinks())));
```

**Pull request merged**